Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb

```diff
@@ -2,119 +2,300 @@
-        if acquire_lock(blob_name + ".lock")
-          return blob_name
+  # Acquire a lease on a blob item with retries.
+  #
+  # By default, it will retry 30 times with 1 second interval.
+  def acquire_lease(blob_name, retry_times = 30, interval_sec = 1)
+    lease = nil;
+    retried = 0;
+    while lease.nil? do
+      begin
+        lease = @azure_blob.acquire_blob_lease(@container, blob_name, {:timeout => 10})
+      rescue StandardError => e
+        if(e.type == 'LeaseAlreadyPresent')
+            if (retried > retry_times)
+                raise
+            end
+            retried += 1
+            sleep interval_sec
         end
       end
-    end
-    return NIL
-  end # def lock_blob
-  
-  def archive_blob(blob_names)
-    real_blob_names = blob_names.select { |name| !name.end_with?(".bak") }
-    real_blob_names.each do |blob_name|
-      new_blob_name = "#{blob_name}-#{Time.now.to_f}.bak"
-      @azure_blob.acquire_lease(@container, blob_name,{:duration=>60, :timeout=>10, :proposed_lease_id=>SecureRandom.uuid})
-      @azure_blob.copy_blob(@container, new_blob_name, @container, blob_name)  
-      @azure_blob.delete_blob(@container, blob_name)
-      return new_blob_name
-    end
-    return NIL
-  end # def lock_blob
- 
-  def process(output_queue)
-    blob_names = list_blob_names
-    
-    if @backupmode
-      blob_name = archive_blob(blob_names)
-    else
-      blob_name = lock_blob(blob_names)
-    end
-
-    return if !blob_name
-    blob, content = @azure_blob.get_blob(@container, blob_name)
-    
-
-    @codec.decode(content) do |event|
-      decorate(event)
-      output_queue << event
-    end
-  rescue LogStash::ShutdownSignal => e
-    raise e
-  rescue => e
-    @logger.error("Oh My, An error occurred.", :exception => e)
-  end # def process
-  
-  public
-  def run(output_queue)
-    while !stop?
-      process(output_queue)
-    end # loop
-  end # def run
- 
-  public
-  def teardown
-  end # def teardown
-end # class LogStash::Inputs::Azuretopic
+    end #while
+    return lease
+  end # acquire_lease
+
+  # Return the next blob for reading as well as the start index.
+  def register_for_read
+    begin
+      all_blobs = list_all_blobs
+      registry = all_blobs.find { |item| item.name.downcase == @registry_path  }
+      registry_locker = all_blobs.find { |item| item.name.downcase == @registry_locker }
+
+      candidate_blobs = all_blobs.select { |item| (item.name.downcase != @registry_path) && ( item.name.downcase != @registry_locker ) }
+      
+      start_index = 0
+      lease = nil;
+
+      # Put lease on locker file than the registy file to allow update of the registry as a workaround for Azure Storage Ruby SDK issue # 16.
+      # Workaround: https://github.com/Azure/azure-storage-ruby/issues/16
+      registry_locker = @azure_blob.create_block_blob(@container, @registry_locker, @reader) if registry_locker.nil?
+      lease = acquire_lease(@registry_locker)
```

Comment:
Again, very good question. Here's the consideration and the implementation (Line 158):
1. acquire_lease is a wrapper of acquire_blob_lease with the retry logic. By default, it will retry 30 times with 1 second interval (30 seconds). This is expected to handle most of the cases.
2. And when it still throws, I think there's something really concern is happening and yes, it throws in that case. One will see it in the logstash self-diagnostics log, with back-trace for trouble shooting.
I agree with you that have a few conflicts from time to time is normal condition. Having 30 conflict in a row is not and I think that should throw.
Does this address the question?